copy/copy.go

```diff
@@ -0,0 +1,308 @@
+	var ci CopyInfo
+	for _, o := range opts {
+		o(&ci)
+	}
+
+	ensureDstPath := dst
+	if d, f := filepath.Split(dst); f != "" && f != "." {
+		ensureDstPath = d
+	}
+	if err := os.MkdirAll(ensureDstPath, 0700); err != nil {
+		return err
+	}
+	dst = filepath.Clean(dst)
+
+	c := newCopier(ci.Chown)
+	srcs := []string{src}
+
+	if ci.AllowWildcards {
+		d1, d2 := splitWildcards(src)
+		if d2 != "" {
+			matches, err := resolveWildcards(d1, d2)
+			if err != nil {
+				return err
+			}
+			srcs = matches
+		}
+	}
+
+	for _, src := range srcs {
+		srcFollowed, err := filepath.EvalSymlinks(src)
+		if err != nil {
+			return err
+		}
+		srcBase := filepath.Base(src)
+		if err := c.copy(ctx, srcBase, srcFollowed, dst); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+type ChownOpt struct {
+	Uid, Gid int
+}
+
+type CopyInfo struct {
+	Chown          *ChownOpt
+	AllowWildcards bool
+}
+
+type Opt func(*CopyInfo)
+
+func WithChown(uid, gid int) Opt {
+	return func(ci *CopyInfo) {
+		ci.Chown = &ChownOpt{Uid: uid, Gid: gid}
+	}
+}
+
+func AllowWildcards(ci *CopyInfo) {
+	ci.AllowWildcards = true
+}
+
+type copier struct {
+	chown  *ChownOpt
+	inodes map[uint64]string
+}
+
+func newCopier(chown *ChownOpt) *copier {
+	return &copier{inodes: map[uint64]string{}, chown: chown}
+}
+
+func (c *copier) copy(ctx context.Context, base, src, dst string) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+	dstBase := base
+	fi, err := os.Lstat(src)
+	if err != nil {
+		return errors.Wrapf(err, "failed to stat %s", src)
+	}
+
+	if _, err := os.Lstat(dst); err != nil {
+		if !os.IsNotExist(err) {
+			return errors.Wrap(err, "failed to lstat destination directory")
+		}
+		dst, dstBase = filepath.Split(dst)
+	}
+
+	target := filepath.Join(dst, dstBase)
+	if !fi.IsDir() {
+		if err := ensureEmptyFileTarget(target); err != nil {
+			return err
+		}
+	}
+
+	switch {
+	case fi.IsDir():
+		if err := c.copyDirectory(ctx, src, target, fi); err != nil {
+			return err
+		}
+	case (fi.Mode() & os.ModeType) == 0:
+		link, err := getLinkSource(target, fi, c.inodes)
+		if err != nil {
+			return errors.Wrap(err, "failed to get hardlink")
+		}
+		if link != "" {
+			if err := os.Link(link, target); err != nil {
+				return errors.Wrap(err, "failed to create hard link")
+			}
+		} else if err := copyFile(src, target); err != nil {
+			return errors.Wrap(err, "failed to copy files")
+		}
+	case (fi.Mode() & os.ModeSymlink) == os.ModeSymlink:
+		link, err := os.Readlink(src)
+		if err != nil {
+			return errors.Wrapf(err, "failed to read link: %s", src)
+		}
+		if err := os.Symlink(link, target); err != nil {
+			return errors.Wrapf(err, "failed to create symlink: %s", target)
+		}
+	case (fi.Mode() & os.ModeDevice) == os.ModeDevice:
+		if err := copyDevice(target, fi); err != nil {
+			return errors.Wrapf(err, "failed to create device")
+		}
+	default:
+		// TODO: Support pipes and sockets
```

Comment:
Is it technically feasible?

Reply:
This is from containerd copyDir source. Pipes should definitely be possible, for sockets could just create empty files.